NXDOC-2956: document Audit Router and Blue/Green Audit migration (#2715)
nuxeo/doc.nuxeo.com-platform-spaces: src/nxdoc/nuxeo-server/data-store/data-modeling/audit/audit-router.md

---
title: Audit Router
description: Route audit log entries to one or more Audit Backends based on event names and predicates.
review:
    comment: ''
    date: '2026-05-07'
    status: ok
labels:
    - audit
    - audit-router
toc: true
tree_item_index: 770
---

{{#> callout type='warning' heading='Available since 2025.16'}}
The Audit Router and the `routes` extension point are introduced in Nuxeo
2025.16. Until then, only the legacy `event` extension point is available.
The legacy `event` extension point still works and remains backward-compatible
on later releases.
{{/callout}}

{{! excerpt}}
The `AuditRouter` is the engine that decides, for each Nuxeo event, which Audit
Backend(s) the resulting `LogEntry` is written to. It is configured through the
`routes` extension point of the Audit Service.
{{! /excerpt}}

## How Routing Works

The audit ingestion pipeline goes through three stages:

1. The `StreamAuditEventListener` intercepts each audited event, asks the
   `AuditRouter` to compute the resulting `LogEntry` from the event context,

>    and appends it to the `audit/audit` Nuxeo Stream.

Reviewer: Is it the listener or the router that appends?

Author: It is the listener.

2. The `audit/writer` computation (`StreamAuditWriter`) consumes the
   `audit/audit` stream entries.

> 3. For every entry, the computation calls back the `AuditRouter` to get the

Reviewer: ```suggestion
3. For every entry, the computation calls the `AuditRouter` back to get the
```

   list of target backends. The router walks the contributed routes and, for
   every route that matches the entry, returns the configured backend. The
   entry is then written to each matching backend.

![]({{file name='architecture-diagram.png' page='audit'}} ?w=600,border=true)

The combination semantics are:

- **routes are ORed**: a single log entry can be written to several backends if
  several routes match.
- **events** of one route are ORed.
- **predicates** of one route are ORed.
- **events AND predicates** are combined inside a single route: a route
  declaring both accepts an entry only when one of its events matches **and**
  one of its predicates is satisfied.

A route may declare only events, only predicates, or both:

> - **Events only** — the events of the route are also registered as auditable,

Reviewer: ```suggestion
- **Events only** — the events declared in the route are registered as auditable event,
```

  so the post-commit listener forwards them to the `audit/audit` stream.
- **Predicates only** — the route does not register events; it only filters
  log entries that already flow through the router (for example, to copy
  entries flowing to the default backend to a secondary one based on their
  category).

## `routes` Extension Point

A route is a `<route>` element with a `name`, a `<backend>` and any
combination of `<event>` and `<predicate>` children:

```xml
<extension target="org.nuxeo.audit.service.AuditComponent" point="routes">
  <route name="default">
    <backend name="default" />
    <event name="documentCreated" />
    <event name="documentModified" enabled="true" />
  </route>
</extension>
```

> 

Reviewer: Would it make sense adding something like: "If no default route is contributed, an implicit one targeting the default backend is created with the platform's auditable events."?

The "implicit `default` route" is mentioned in some places, but sounds it could be stated here.

Author: I would add something like _By default Nuxeo Platform contributes the `default` route holding all auditable events_

- Setting `enabled="false"` is the supported way to suppress an event from a
  route without redefining it; this is how the SQL backend disables `search`
  by default.

### `<predicate>`

`<predicate class="..."/>` filters which log entries the route accepts. A
predicate is a `java.util.function.Predicate<LogEntry>` that takes a
`Map<String, String>` of `<property>` entries as constructor argument:

```xml
<route name="documentCategory">
  <backend name="default" />
  <predicate class="org.nuxeo.audit.service.route.CategoryLogEntryPredicate">
    <property name="category">eventDocumentCategory</property>
  </predicate>
</route>
```

The platform ships `CategoryLogEntryPredicate`, which keeps log entries whose
`category` matches the configured value. You can contribute your own
`Predicate<LogEntry>` implementation by providing a public constructor that
accepts a `Map<String, String>`.

### Route Inheritance with `copy`

A route can copy another existing route by referring to it via the `copy`
attribute. The copying route inherits the backend and the events/predicates
of the source route, and overrides them with what it declares locally:

> 

Reviewer: Claude suggests the following, I'm wondering 🤔 :

> "Overrides" is ambiguous: does a locally declared `<event>` replace the inherited list of events (8c1424e6acd61385ab834a0ab67dccecebe319dd's SQL package uses `<event name="search" enabled="false"/>` to disable a single event, implying merge), or does the local declaration completely replace the inherited list
>
> Please clarify, e.g.:                                                                                                                                                              
> "Locally declared `<event>` and `<predicate>` elements are merged with those of the copied route. A local `<event name="..." enabled="false"/>` disables a specific inherited event. The `<backend>` is replaced when redeclared."

Author: We can rephrase yes, by default we tend to have mergeable list/map in descriptor if they have an id (which is the case for `event@name` and `predicate@name`). We should probably use _merge_ instead of override_.

---
title: Audit Router
description: Route audit log entries to one or more Audit Backends based on event names and predicates.
review:
    comment: ''
    date: '2026-05-07'
    status: ok
labels:
    - audit
    - audit-router
toc: true
tree_item_index: 770
---

{{#> callout type='warning' heading='Available since 2025.16'}}
The Audit Router and the `routes` extension point are introduced in Nuxeo
2025.16. Until then, only the legacy `event` extension point is available.
The legacy `event` extension point still works and remains backward-compatible
on later releases.
{{/callout}}

{{! excerpt}}
The `AuditRouter` is the engine that decides, for each Nuxeo event, which Audit
Backend(s) the resulting `LogEntry` is written to. It is configured through the
`routes` extension point of the Audit Service.
{{! /excerpt}}

## How Routing Works

The audit ingestion pipeline goes through three stages:

1. The `StreamAuditEventListener` intercepts each audited event, asks the
   `AuditRouter` to compute the resulting `LogEntry` from the event context,

>    and the listener appends it to the `audit/audit` Nuxeo Stream.

Reviewer 2: ```suggestion
   The `StreamAuditEventListener` intercepts each audited event, asks the `AuditRouter` to compute the resulting `LogEntry` from the event context, and appends it to the `audit/audit` Nuxeo Stream."
```

2. The `audit/writer` computation (`StreamAuditWriter`) consumes the
   `audit/audit` stream entries.
3. For every entry, the computation calls the `AuditRouter` back to get the
   list of target backends. The router walks the contributed routes and, for
   every route that matches the entry, returns the configured backend. The
   entry is then written to each matching backend.

![]({{file name='architecture-diagram.png' page='audit'}} ?w=600,border=true)

The combination semantics are:

- **routes are ORed**: a single log entry can be written to several backends if
  several routes match.
- **events** of one route are ORed.
- **predicates** of one route are ORed.
- **events AND predicates** are combined inside a single route: a route
  declaring both accepts an entry only when one of its events matches **and**
  one of its predicates is satisfied.

A route may declare only events, only predicates, or both:

> - **Events only** — the events declared in the route are registered as auditable event,

Reviewer 2: ```suggestion
- **Events only** — the events declared in the route are registered as auditable events,
```

---
title: Audit Router
description: Route audit log entries to one or more Audit Backends based on event names and predicates.
review:
    comment: ''
    date: '2026-05-07'
    status: ok
labels:
    - audit
    - audit-router
toc: true
tree_item_index: 770
---

{{#> callout type='warning' heading='Available since 2025.16'}}
The Audit Router and the `routes` extension point are introduced in Nuxeo 2025.16.
On previous releases the legacy `event` extension point still works and remains

> backward-compatible.

Author: Isn't weird? I understand it like the event extension point only works on 2025.16 despite the fact we're talking about backward-compatible, could you re-phrase?

{{/callout}}

{{! excerpt}}
The `AuditRouter` is the engine that decides, for each Nuxeo event, which Audit
Backend(s) the resulting `LogEntry` is written to. It is configured through the
`routes` extension point of the Audit Service.
{{! /excerpt}}

## How Routing Works

When an audited event reaches the `audit/writer` stream computation, the
`AuditRouter`:

> 1. Computes the candidate `LogEntry` from the event.

Author: This is not entirely true, the LogEntry from the event is done by the AuditRouter within the `StreamAuditEventListener`, then inserted in the `audit/audit` stream that is consumed by the `audit/writer` computation and then point 2. like you state.

I think it's worth to insert a diagram here, could you draw a proposition? You may be interested within the _nuxeo-audit-architecture-simple-routing_ diagram on https://hyland.atlassian.net/wiki/spaces/NuxEng/pages/3837330155/Architecture+Multiple+Audit+Backends page. (If you draw a draw.io one, which sounds like a good way, please update the one on the confluence as we can keep track of it and update it later, and do an export for the documentation repository).

2. Walks the contributed routes.
3. For every route whose **events** AND **predicates** match the entry, writes
   it to the configured backend.

> The combination semantics are:
> 
> - **routes are ORed**: a single log entry can be written to several backends if
>   several routes match.
> - **events** of one route are ORed.
> - **predicates** of one route are ANDed.
> - **events AND predicates** are combined inside a single route, so a route
>   accepts an entry when one of its events matches **and** every predicate is
>   satisfied.
> 
> A route may declare only events, only predicates, or both:
> 
> - **Events only** — the events of the route are also registered as auditable,
>   so the post-commit listener forwards them to the `audit/audit` stream.
> - **Predicates only** — the route does not register events; it only filters
>   log entries that already flow through the router (for example, to copy
>   entries flowing to the default backend to a secondary one based on their
>   category).

Author: ```suggestion
The combination semantics are:

- **routes are ORed**: a single log entry can be written to several backends if
  several routes match.
- **events** of one route are ORed.
- **predicates** of one route are ORed.
- **events AND predicates** are combined inside a single route, so a route
  accepts an entry when one of its events matches **and** any predicate is
  satisfied.

A route may declare only events, only predicates, or both:

- **Events only** — the events of the route are also registered as auditable,
  so the post-commit listener forwards them to the `audit/audit` stream.
- **Predicates only** — the route does not register events; it only filters
  log entries that already flow through the router (for example, to copy
  entries flowing to the default backend to a secondary one based on their
  category).
```

It may worth to check other content about this mistake on ORed predicates.

</extension>
```

### 3. Configure the OpenSearch client used by the secondary backend

```xml
<extension target="org.nuxeo.runtime.opensearch1.OpenSearchComponent" point="client">
  <client id="audit/secondary">
    <server>${hyland.custom.audit.backend.secondary.opensearch1.client.server:=http://localhost:9200}</server>
    <connectionTimeout>${hyland.custom.audit.backend.secondary.opensearch1.client.connectionTimeout:=30s}</connectionTimeout>
    <socketTimeout>${hyland.custom.audit.backend.secondary.opensearch1.client.socketTimeout:=121000ms}</socketTimeout>
    <username>${hyland.custom.audit.backend.secondary.opensearch1.client.username:=}</username>
    <password>${hyland.custom.audit.backend.secondary.opensearch1.client.password:=}</password>
  </client>
</extension>
```

### 4. Contribute the route

```xml
<extension target="org.nuxeo.audit.service.AuditComponent" point="routes">
  <route name="secondary_route">
    <backend name="secondary" />
    <event name="hylandBusinessEvent" />
  </route>
</extension>
```

From now on, every `hylandBusinessEvent` raised by the application is also
ingested by the secondary OpenSearch backend, while regular events keep

> flowing to the default backend.

Author: It is not very clear that the hylandBusinessEvent only goes to secondary backend and not the default one, please rephrase.

> {{#> callout type='note' }}

Author: What about info or warn? sounds like a big thing.

When installing two Audit Backend Marketplace Packages side-by-side, only one
of them must be the default. Force the default factory in `nuxeo.conf` and
disable the other one, for instance:

```
# default backend is SQL
nuxeo.audit.backend.default.factory=org.nuxeo.audit.sql.SQLAuditBackendFactory
# secondary OpenSearch package must not register itself as default
nuxeo.audit.backend.default.opensearch1.enabled=false
```
{{/callout}}

## Backward Compatibility with `event`

The legacy `event` extension point is still honored. Any contribution made on
`event` is automatically merged into the implicit `default` route during
component startup. Migrating to the `routes` extension point is recommended

> but not required to upgrade.

Author: The end of the sentence _Migrating to the `routes` extension point is recommended but not required to upgrade._ is not totally true, this is not a mandatory step, but still recommended as former extension point will be removed in the future.

For details about how to migrate existing contributions, see the
[How to Upgrade Nuxeo Audit Service]({{page page='how-to-upgrade-audit-service'}})

> guide.

Author: This is not linked to the router, this is linked to the general topic, maybe this information would make sense in the main page, but not here.

Author: Not taken into account, please delete this sentence.

## Introspection

Use `GET /nuxeo/api/v1/management/audit/introspection` to retrieve a PlantUML
representation of the live audit routing topology. See the

> [Audit Endpoint](https://doc.nuxeo.com/rest-api/1/audit-endpoint/) for details.

Author: Good! Can't we do a reference with `{{page...}}`? I think it would be better, try to find another link to rest-api documentation in this repository.

---
title: Audit Router
description: Route audit log entries to one or more Audit Backends based on event names and predicates.
review:
    comment: ''
    date: '2026-05-07'
    status: ok
labels:
    - audit
    - audit-router
toc: true
tree_item_index: 770
---

{{#> callout type='warning' heading='Available since 2025.16'}}
The Audit Router and the `routes` extension point are introduced in Nuxeo
2025.16. Until then, only the legacy `event` extension point is available.
The legacy `event` extension point still works and remains backward-compatible
on later releases.
{{/callout}}

{{! excerpt}}
The `AuditRouter` is the engine that decides, for each Nuxeo event, which Audit
Backend(s) the resulting `LogEntry` is written to. It is configured through the
`routes` extension point of the Audit Service.
{{! /excerpt}}

## How Routing Works

The audit ingestion pipeline goes through three stages:

> 1. The `StreamAuditEventListener` (synchronous, priority `500`) intercepts

Author: Don't talk about priority, it is low level stuff!

   each audited event, asks the `AuditRouter` to compute the resulting
   `LogEntry` from the event context, and appends it to the `audit/audit`
   Nuxeo Stream.
2. The `audit/writer` computation (`StreamAuditWriter`) consumes the
   `audit/audit` stream entries.
3. For every entry, the computation calls back the `AuditRouter` to get the
   list of target backends. The router walks the contributed routes and, for
   every route whose **events** OR **predicates** match the entry, returns
   the configured backend. The entry is then written to each matching

>    backend.

Author: event are ANDed with predicates, not sure about the OR you set?
